File: dotnet-desktop-guide/wpf/events/snippets/attached-events-overview/WpfControlLibraryVb/AquariumFilter.vb
Imports System.Windows

'<AddRemoveHandlersForAttachedEvent>
Public Class AquariumFilter

    ' Register a custom routed event using the bubble routing strategy.
    Public Shared ReadOnly CleanEvent As RoutedEvent = EventManager.RegisterRoutedEvent(
        "Clean", RoutingStrategy.Bubble, GetType(RoutedEventHandler), GetType(AquariumFilter))

    ' Provide an add handler accessor method for the Clean event.
    Public Shared Sub AddCleanHandler(dependencyObject As DependencyObject, handler As RoutedEventHandler)
        Dim uiElement As UIElement = TryCast(dependencyObject, UIElement)

        If uiElement IsNot Nothing Then
            uiElement.[AddHandler](CleanEvent, handler)
        End If
    End Sub

    ' Provide a remove handler accessor method for the Clean event.
    Public Shared Sub RemoveCleanHandler(dependencyObject As DependencyObject, handler As RoutedEventHandler)
        Dim uiElement As UIElement = TryCast(dependencyObject, UIElement)

        If uiElement IsNot Nothing Then
            uiElement.[RemoveHandler](CleanEvent, handler)
        End If
    End Sub

End Class
'</AddRemoveHandlersForAttachedEvent>

File: dotnet-desktop-guide/wpf/events/snippets/attached-events-overview/WpfControlLibraryCsharp/AquariumFilter.cs
using System.Windows;

namespace WpfControlCsharp
{
    //<AddRemoveHandlersForAttachedEvent>
    public class AquariumFilter
    {
        // Register a custom routed event using the bubble routing strategy.
        public static readonly RoutedEvent CleanEvent = EventManager.RegisterRoutedEvent(
            "Clean", RoutingStrategy.Bubble, typeof(RoutedEventHandler), typeof(AquariumFilter));

        // Provide an add handler accessor method for the Clean event.
        public static void AddCleanHandler(DependencyObject dependencyObject, RoutedEventHandler handler)
        {
            if (dependencyObject is not UIElement uiElement)
                return;

            uiElement.AddHandler(CleanEvent, handler);
        }

        // Provide a remove handler accessor method for the Clean event.
        public static void RemoveCleanHandler(DependencyObject dependencyObject, RoutedEventHandler handler)
        {
            if (dependencyObject is not UIElement uiElement)
                return;

            uiElement.RemoveHandler(CleanEvent, handler);
        }
    }
    //</AddRemoveHandlersForAttachedEvent>
}
